### PR TITLE
sql/pgwire: tidy up tests for `Exec` with placeholders

### PR DESCRIPTION
**Pull request by tamird**
[Review on Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/3856)

### CODE REVIEW

File: sql/pgwire_test.go [hunk 1]
```diff
@@ -325,3 +324,83 @@ func TestPGPrepared(t *testing.T) {
...
+		}
+
+		for _, test := range tests {
+			var results []result
+			{
+				res, err := db.Exec(query, test.params...)
+				results = append(results, result{res: res, err: err})
+			}
+			{
+				res, err := stmt.Exec(test.params...)
```
Comment (petermattis) [inline sql/pgwire_test.go]: What is this about?

**Comment by petermattis** [comment]
LGTM

**Comment by tamird** [comment]
This is related to #3819.

**Comment by RaduBerinde** [comment]
*[sql/pgwire_test.go, line 363 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3856#-K81LxLPNyH9myBjU5a5:-K81LxLPNyH9myBjU5a6:161717605) ([raw file](https://github.com/cockroachdb/cockroach/blob/89fce8a333f22cff39e404d788df191deb2d2514/sql/pgwire_test.go#L363)):*
Wow.. This is kind of insane. Can't we do something more readable, e.g.
```
for i := 0; i < 2; i++) {
   if (i == 0) {
      res, err := db.Exec()
   } else {
      res, err := stmt.Exec()
   }
   .. check error ..
```
Actually, it seems that the errors are supposed to be the same (they both must match test.error right?). I would just run both statements and first compare the errors to check they are the same. And then do the error checking once.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/3856)*

**Comment by tamird** [comment]
Review status: 0 of 1 files reviewed at latest revision, 1 unresolved discussion, some commit checks failed.

---

*[sql/pgwire_test.go, line 363 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3856#-K81LxLPNyH9myBjU5a5:-K853CXyYE20P9wjggUB:-1010924483) ([raw file](https://github.com/cockroachdb/cockroach/blob/89fce8a333f22cff39e404d788df191deb2d2514/sql/pgwire_test.go#L363)):*
I think this pattern is quite idiomatic in Go. checking if both errors are identical first can lead to some hard to debug error messages; this version produces a more self contained error in case this test fails. One problem is that it'll be hard to tell which variant failed (stmt.Exec or db.Exec) - I'll see about improving that.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/3856)*

**Comment by bdarnell** [comment]
Review status: 0 of 1 files reviewed at latest revision, 1 unresolved discussion, some commit checks failed.

---

*[sql/pgwire_test.go, line 363 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3856#-K81LxLPNyH9myBjU5a5:-K85HrO7k70_Z1y5cujU:1927492681) ([raw file](https://github.com/cockroachdb/cockroach/blob/89fce8a333f22cff39e404d788df191deb2d2514/sql/pgwire_test.go#L363)):*
I'm not sure what you're referring to by "this pattern", but I wouldn't consider defining functions inline in a `for` statement to be idiomatic. I prefer Radu's version looping from 0..2. (the error checking is fine as-is)

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/3856)*

**Comment by RaduBerinde** [comment]
*[sql/pgwire_test.go, line 363 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3856#-K81LxLPNyH9myBjU5a5:-K85R6G6zhWq4rjfWDNi:1842565877) ([raw file](https://github.com/cockroachdb/cockroach/blob/89fce8a333f22cff39e404d788df191deb2d2514/sql/pgwire_test.go#L363)):*
You can also separate the three ifs into a function (which might be useful for other cases as well) and just call it twice (no loop).

```
res, err := db.Exec()
checkErrorMatches(err, test.error);
res, err = stmt.Exec()
checkErrorMatches(err, test.error)

```

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/3856)*

**Comment by nvanbenschoten** [comment]
Review status: 0 of 1 files reviewed at latest revision, 3 unresolved discussions.

---

*[sql/pgwire_test.go, line 346 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3856#-K9Sj4fi8SA3dEXGTy3l:-K9Sj4fi8SA3dEXGTy3m:1191475638) ([raw file](https://github.com/cockroachdb/cockroach/blob/e39998ffebccf06e005352df2bb8dd01f40ef1ae/sql/pgwire_test.go#L346)):*
You might want to `s/queryTests/execTests/g`, and `s/query/exec/g` below.

---

*[sql/pgwire_test.go, line 383 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3856#-K80ZjrwYnT6yuFVuRMI-r1-383:-K9SkzH88KfrcmOVE1m_:1805387003) ([raw file](https://github.com/cockroachdb/cockroach/blob/e39998ffebccf06e005352df2bb8dd01f40ef1ae/sql/pgwire_test.go#L383)):*
You're not defering anything, so why is this needed?

---

*[sql/pgwire_test.go, line 363 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3856#-K81LxLPNyH9myBjU5a5:-K9SlImbeAprsGrNHq3t:-1627556861) ([raw file](https://github.com/cockroachdb/cockroach/blob/89fce8a333f22cff39e404d788df191deb2d2514/sql/pgwire_test.go#L363)):*
Personally I'd be ok with this approach if the function slice wasn't defined inline, but instead assigned to a well-named variable. It seems like we're doing the same thing in `TestPGPreparedQuery`, so whatever we decide, I'd make sure to keep them consistent.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/3856)*

**Comment by petermattis** [comment]
LGTM modulo some of the other comments below.

---

Review status: 0 of 1 files reviewed at latest revision, 4 unresolved discussions.

---

*[sql/pgwire_test.go, line 262 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3856#-K9SnG14Kcy2tibjP7DX:-K9SnG14Kcy2tibjP7DY:-486250862) ([raw file](https://github.com/cockroachdb/cockroach/blob/dbcd932904f11a05e6489d2f3511563fd50ebbd0/sql/pgwire_test.go#L262)):*
Not sure the extra indentation is worth being able to `defer rows.Close()`, but whatever.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/3856#-:-K9SnnbY_AsW0UGkmwmT:-1252198838)*

**Comment by tamird** [comment]
PTAL, this got a bit bigger.

---

Review status: 0 of 1 files reviewed at latest revision, 4 unresolved discussions.

---

*[sql/pgwire_test.go, line 346 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3856#-K9Sj4fi8SA3dEXGTy3l:-K9SmIjLM-0vub2fSdOW:1636242043) ([raw file](https://github.com/cockroachdb/cockroach/blob/e39998ffebccf06e005352df2bb8dd01f40ef1ae/sql/pgwire_test.go#L346)):*
changed queryTests, but `query` is still appropriate since it's a string.

---

*[sql/pgwire_test.go, line 383 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3856#-K80ZjrwYnT6yuFVuRMI-r1-383:-K9SmcMPPte2ZP1w97x3:1740281830) ([raw file](https://github.com/cockroachdb/cockroach/blob/e39998ffebccf06e005352df2bb8dd01f40ef1ae/sql/pgwire_test.go#L383)):*
Removed!

---

*[sql/pgwire_test.go, line 363 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3856#-K81LxLPNyH9myBjU5a5:-K9SmWrGvRFfxPnORikV:2071709664) ([raw file](https://github.com/cockroachdb/cockroach/blob/89fce8a333f22cff39e404d788df191deb2d2514/sql/pgwire_test.go#L363)):*
I don't really see the big problem here; this inline function is just currying `db.Exec`

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/3856#-:-K9TSVQ0CyZnxUwHlvgM:-1531968614)*

**Comment by mjibson** [comment]
LGTM

---

Review status: 0 of 1 files reviewed at latest revision, 3 unresolved discussions.

---



*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/3856#-:-K9YXU2jS7CehS6hbY5C:676834233)*

**Pull request merged**